### ISSUE TITLE
Segmentation fault while compiling a program including a non-existant derived type.

### ISSUE THREAD

**Issue by p-pap** [lfortran/lfortran#2944]
Consider the following program (LFortran 0.29.0):
```
program test
implicit none
type(foo) :: c
end program
```
I would expect an error saying something like "derived type `foo` is used before defined". Instead, I get a segmentation fault. In this simple program, it's easy to figure out what's going on, but in larger programs it won't.

**Comment by certik** [comment]
Thanks, indeed that's a bug. Here is what I get with LFortran built in Debug mode:
```
$ lfortran a.f90 
ASR verify pass error: ASR verify: The symbol table was found, but the symbol `sym` is not in it
 --> a.f90:3:1
  |
3 | type(foo) :: c
  | ^^^^^^^^^^^^^^ failed here

ASR verify pass error: ASR verify: Struct::m_derived_type 'foo' cannot point outside of its symbol table, owner: test
 --> a.f90:3:1
  |
3 | type(foo) :: c
  | ^^^^^^^^^^^^^^ failed here
```
It's an internal bug. The frontend should report a nice error message to the user, as you said.

**Comment by p-pap** [comment]
Just to clarify, shall I use a Debug build for reporting issues? I thought Release is more appropriate because that is what people will probably use.

**Comment by certik** [comment]
No, just use Release mode for reporting. Debug mode is meant for developers only, it does many additional checks, but those are all turned off in Release mode, which is made for speed of compilation. It won't segfault once we iron out these bugs.
